File: backend_py/mcp.py
"""MCP client: built-in tools (echo, add, Google Calendar, Google Docs) + external MCP servers."""
from store.connectors import get_connector
from google_calendar import create_calendar_event, list_calendar_events
from google_docs import create_google_doc

try:
    from external_mcp import list_external_tools, call_external_tool, has_external_tool
except ImportError:
    def list_external_tools(user_id): return []
    def call_external_tool(user_id, name, args): return f"Tool {name} not available."
    def has_external_tool(user_id, name): return False

CALENDAR_TOOLS = [
    {"name": "create_calendar_event", "description": "Create a new event on the user's Google Calendar. Pass start and end in ISO 8601 UTC.", "inputSchema": {"type": "object", "properties": {"summary": {"type": "string"}, "startTime": {"type": "string"}, "endTime": {"type": "string"}, "description": {"type": "string"}}, "required": ["summary", "startTime", "endTime"]}},
    {"name": "list_calendar_events", "description": "List upcoming events from the user's Google Calendar.", "inputSchema": {"type": "object", "properties": {"maxResults": {"type": "number"}}}},
    {"name": "create_google_doc", "description": "Create a new Google Doc with the given title and body content.", "inputSchema": {"type": "object", "properties": {"title": {"type": "string"}, "content": {"type": "string"}}, "required": ["title", "content"]}},
]
MOCK_TOOLS = [
    {"name": "echo", "description": "Echo back the message", "inputSchema": {"type": "object", "properties": {"message": {"type": "string"}}}},
    {"name": "add", "description": "Add two numbers", "inputSchema": {"type": "object", "properties": {"a": {"type": "number"}, "b": {"type": "number"}}}},
]

def create_mcp_client(user_id: str | None):
    if not user_id:
        return UserMcpClient("", None)
    conn = get_connector(user_id, "google")
    creds = conn.get("credentials") if conn else None
    return UserMcpClient(user_id, creds)

class UserMcpClient:
    def __init__(self, user_id: str, calendar_credentials: str | None):
        self.user_id = user_id
        self.calendar_credentials = calendar_credentials

    def list_tools(self) -> list[dict]:
        tools = list(MOCK_TOOLS)
        if self.calendar_credentials:
            tools = tools + list(CALENDAR_TOOLS)
        built_in_names = {t["name"] for t in tools}
        for t in list_external_tools(self.user_id):
            if t.get("name") not in built_in_names:
                tools.append(t)
                built_in_names.add(t.get("name"))
        return tools

    def call_tool(self, name: str, args: dict) -> str:
        if name == "echo":
            return str(args.get("message", ""))
        if name == "add":
            return str(float(args.get("a", 0)) + float(args.get("b", 0)))
        if name == "create_calendar_event" and self.calendar_credentials:
            summary = str(args.get("summary", ""))
            start_time = str(args.get("startTime", ""))
            end_time = str(args.get("endTime", ""))
            desc = args.get("description")
            if not summary or not start_time or not end_time:
                return "Error: summary, startTime, and endTime are required."
            try:
                return create_calendar_event(self.calendar_credentials, summary, start_time, end_time, desc)
            except Exception as e:
                return f"Error creating event: {e}"
        if name == "list_calendar_events" and self.calendar_credentials:
            try:
                return list_calendar_events(self.calendar_credentials, int(args.get("maxResults", 10)))
            except Exception as e:
                return f"Error listing events: {e}"
        if name == "create_google_doc" and self.calendar_credentials:
            title = str(args.get("title", "")).strip()
            content = str(args.get("content", "")).strip()
            if not title or not content:
                return "Error: title and content are required for create_google_doc."
            try:
                return create_google_doc(self.calendar_credentials, title, content)
            except Exception as e:
                return f"Error creating document: {e}"
        if has_external_tool(self.user_id, name):
            return call_external_tool(self.user_id, name, args)
        return f'Tool "{name}" is not available. Connect Google or add external MCP servers.'

File: backend/src/mcp.ts
/**
 * MCP client: list tools and call tools.
 * Built-in: Google (Calendar, Docs), echo, add.
 * External: from MCP_SERVERS_JSON (Brave Search, Time, Todo, etc.) — see docs/EXTERNAL_MCP_FOR_STUDENTS.md.
 */

import { getConnector } from './store/connectors.js';
import { createCalendarEvent, listCalendarEvents } from './calendar.js';
import { createGoogleDoc } from './google-docs.js';
import { listExternalTools, callExternalTool, hasExternalTool } from './external-mcp.js';

export interface McpTool {
  name: string;
  description?: string;
  inputSchema?: Record<string, unknown>;
}

export interface McpClientInterface {
  listTools(): Promise<McpTool[]>;
  callTool(name: string, args: Record<string, unknown>): Promise<string>;
}

const CALENDAR_TOOLS: McpTool[] = [
  {
    name: 'create_calendar_event',
    description: 'Create a new event on the user\'s Google Calendar. Pass start and end in ISO 8601 UTC (e.g. 2026-03-13T16:55:00Z). The server converts to the user\'s timezone. If the user does not specify duration, use 1 hour. endTime must be after startTime.',
    inputSchema: {
      type: 'object',
      properties: {
        summary: { type: 'string', description: 'Event title' },
        startTime: { type: 'string', description: 'Start datetime ISO 8601 UTC (e.g. 2026-03-13T16:55:00Z)' },
        endTime: { type: 'string', description: 'End datetime ISO 8601 UTC; must be after startTime' },
        description: { type: 'string', description: 'Optional event description' },
      },
      required: ['summary', 'startTime', 'endTime'],
    },
  },
  {
    name: 'list_calendar_events',
    description: 'List upcoming events from the user\'s Google Calendar.',
    inputSchema: {
      type: 'object',
      properties: {
        maxResults: { type: 'number', description: 'Max events to return (default 10)' },
      },
    },
  },
  {
    name: 'create_google_doc',
    description: 'Create a new Google Doc with the given title and body content. Use this when the user asks to summarize a page and put it in a doc, or to create a document with specific content. The content is the full body text (e.g. a summary).',
    inputSchema: {
      type: 'object',
      properties: {
        title: { type: 'string', description: 'Document title (e.g. "Research Summary")' },
        content: { type: 'string', description: 'Full body text to put in the document' },
      },
      required: ['title', 'content'],
    },
  },
];

const MOCK_TOOLS: McpTool[] = [
  { name: 'echo', description: 'Echo back the message', inputSchema: { type: 'object', properties: { message: { type: 'string' } } } },
  { name: 'add', description: 'Add two numbers', inputSchema: { type: 'object', properties: { a: { type: 'number' }, b: { type: 'number' } } } },
];

export class UserMcpClient implements McpClientInterface {
  constructor(
    private userId: string,
    private calendarCredentials: string | null
  ) {}

  async listTools(): Promise<McpTool[]> {
    const tools = [...MOCK_TOOLS];
    if (this.calendarCredentials) tools.push(...CALENDAR_TOOLS);
    const builtInNames = new Set(tools.map((t) => t.name));
    const external = await listExternalTools(this.userId);
    for (const t of external) {
      if (!builtInNames.has(t.name)) {
        tools.push(t);
        builtInNames.add(t.name);
      }
    }
    return tools;
  }

  async callTool(name: string, args: Record<string, unknown>): Promise<string> {
    if (name === 'echo') return String(args.message ?? '');
    if (name === 'add') return String(Number(args.a) + Number(args.b));

    if (name === 'create_calendar_event' && this.calendarCredentials) {
      const summary = String(args.summary ?? '');
      const startTime = String(args.startTime ?? '');
      const endTime = String(args.endTime ?? '');
      const description = args.description != null ? String(args.description) : undefined;
      if (!summary || !startTime || !endTime) return 'Error: summary, startTime, and endTime are required.';
      try {
        return await createCalendarEvent(this.calendarCredentials, summary, startTime, endTime, description);
      } catch (e) {
        return `Error creating event: ${e instanceof Error ? e.message : String(e)}`;
      }
    }

    if (name === 'list_calendar_events' && this.calendarCredentials) {
      const maxResults = typeof args.maxResults === 'number' ? args.maxResults : 10;
      try {
        return await listCalendarEvents(this.calendarCredentials, maxResults);
      } catch (e) {
        return `Error listing events: ${e instanceof Error ? e.message : String(e)}`;
      }
    }

    if (name === 'create_google_doc' && this.calendarCredentials) {
      const title = String(args.title ?? '').trim();
      const content = String(args.content ?? '').trim();
      if (!title || !content) return 'Error: title and content are required for create_google_doc.';
      try {
        return await createGoogleDoc(this.calendarCredentials, title, content);
      } catch (e) {
        return `Error creating document: ${e instanceof Error ? e.message : String(e)}`;
      }
    }

    const externalHas = await hasExternalTool(this.userId, name);
    if (externalHas) return callExternalTool(this.userId, name, args);

    return `Tool "${name}" is not available. Connect Google in Connectors for Calendar/Docs, or add external MCP servers via MCP_SERVERS_JSON.`;
  }
}

export async function createMcpClient(userId?: string | null): Promise<McpClientInterface> {
  if (!userId) {
    return new UserMcpClient('', null);
  }
  const googleConn = await getConnector(userId, 'google');
  const googleCredentials = googleConn?.credentials ?? null;
  return new UserMcpClient(userId, googleCredentials);
}
